node/pkg/governor/governor_test.go

```diff
@@ -2808,6 +2808,100 @@ func TestCoinGeckoQueries(t *testing.T) {
 	}
 }
 
+// Test the URL of CoinGecko queries to be correct
+func TestCoinGeckoQueryFormat(t *testing.T) {
+	id_amount := 10
+	ids := make([]string, id_amount)
+	for idx := 0; idx < id_amount; idx++ {
+		ids[idx] = fmt.Sprintf("id%d", idx)
+	}
+
+	// Create and parse the query
+	queries := createCoinGeckoQueries(ids, 100, "") // Not API key
+	require.Equal(t, len(queries), 1)
+	query_url, err := url.Parse(queries[0])
+	require.Equal(t, err, nil)
+	params, err := url.ParseQuery(query_url.RawQuery)
+	require.Equal(t, err, nil)
+
+	// Test the portions of the URL for the non-pro version of the API
+	require.Equal(t, query_url.Scheme, "https")
+	require.Equal(t, query_url.Host, "api.coingecko.com")
+	require.Equal(t, query_url.Path, "/api/v3/simple/price")
+	require.Equal(t, params.Has("x_cg_pro_api_key"), false)
+	require.Equal(t, params.Has("vs_currencies"), true)
+	require.Equal(t, params["vs_currencies"][0], "usd")
+	require.Equal(t, params.Has("ids"), true)
+
+	// Create and parse the query with an API key
+	queries = createCoinGeckoQueries(ids, 100, "FAKE_KEY") // With API key
+	require.Equal(t, len(queries), 1)
+	query_url, err = url.Parse(queries[0])
+	require.Equal(t, err, nil)
+	params, err = url.ParseQuery(query_url.RawQuery)
+	require.Equal(t, err, nil)
+
+	// Test the portions of the URL actually provided
+	require.Equal(t, query_url.Scheme, "https")
+	require.Equal(t, query_url.Host, "pro-api.coingecko.com")
+	require.Equal(t, query_url.Path, "/api/v3/simple/price")
+	require.Equal(t, params.Has("x_cg_pro_api_key"), true)
+	require.Equal(t, params["x_cg_pro_api_key"][0], "FAKE_KEY")
+	require.Equal(t, params.Has("vs_currencies"), true)
+	require.Equal(t, params["vs_currencies"][0], "usd")
+	require.Equal(t, params.Has("ids"), true)
+
+}
+
+// Test that the prices are gathering from the CoinGecko API
+func TestCoinGeckoPriceChecks(t *testing.T) {
```

Comment:
When I originally wrote the stuff to query Coin Gecko, I suggested writing tests to do queries, and some people (@SEJeff  as I recall) were adamant that go tests should not require access to the internet. That's why I went the `check_query.go` route instead. Not sure if that feeling has changed.

Reply:
👍🏻 I think connecting to the internet gets into integration test territory. Maybe Tilt tests would be more appropriate?

Reply:
Not testing this functionality at all in the codebase seems like a mistake waiting to happen tbh. What's wrong with connecting to the Internet on this test?

Reply:
It's kind of a philosophical thing but a lot of people have the view that unit tests should strictly test your own functions/things strictly under your control/only your software and not interactions with other software.

So in this case that interaction is with the network, but similarly an author might avoid interacting with a database or the file system. 

In these circumstances the unit tests aren't very unit-y. Here we would be testing 1) parsing JSON into Go structs but also 2) our connectivity to CoinGecko. If CoinGecko is down, should this make our build fail? If a developer is running tests against an unrelated feature, how will they know whether they broke logic or whether this test failed due to internet connectivity issues? What if CoingGecko changes their API interface or the format of their responses? We might not want changes in their software to break our software development pipeline. 

We might also want to run the unit tests in a restricted environment that doesn't have general outgoing network access. 

I think if you want to test this you could try to mock it: in-line JSON of the right format here and make sure that the data structures get parsed correctly.
You could also try breaking the functions into smaller pieces. You isolate the network request to a function that's a single line long, and then unit test every thing else around it so you can be sure that part is working fine.

Reply:
> What if CoingGecko changes their API interface or the format of their responses?

This is exactly what I think would be great to test actually. Would mean we catch any breaking changes early. If there's a way to differentiate using the HTTP error codes between CoinGecko being down and any other issues, that could alleviate some of the downside of having an internet connected test

Reply:
I see the point. It may make the testing unstable. If the API bans the IP testing or is having issues resolving for whatever reason then it would cause the test to fail for no reason. Hmmmm. 

If the API interface breaks for CoinGecko, we would WANT to know that in our tests I would think; I would agree with Dirk on that. Additionally, I want people who are touching this code to make sure that the endpoint is correct and configured properly when the code is modified. 

> I think if you want to test this you could try to mock it: in-line JSON of the right format here and make sure that the data structures get parsed correctly.

I feel this would test SOME of the code paths but not all of them. This is why I wrote the tests this may.

Reply:
@bruce-riley Here are the options that I see:

- Leave the test as is with the network requests
- Change the test to *mock* the response in some way. This would miss large parts I wanted to test though.
- Comment out the test but keep it for people who want to use it. 
- Remove the test altogether. 

What are your thoughts?

Reply:
> Not testing this functionality at all in the codebase seems like a mistake waiting to happen tbh. What's wrong with connecting to the Internet on this test?

Unit tests and for testing controlled environments. Relying on an external web service is NOT a unit test. This is the wrong way to do software engineering. You can mockup an example coingecko server with fixture json returned with the [httptest](https://pkg.go.dev/net/http/httptest) package. @bruce-riley is correct here.

Reply:
From discussions on another channel, I think Jeff and Bruce are right. I removed this test but plan on making a Github Actions PR cron job instead. I'll. be making that in a separate PR though. Thanks all!